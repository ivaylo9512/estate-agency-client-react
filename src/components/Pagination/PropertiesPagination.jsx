import { useEffect } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { useState } from "react";

const PropertiesPagination = ({selector, setCurrentProperties, getProperties, pagesPerSlide = 4}) => {
    const {dataInfo: { data, maxPages, pages, isLoading }, query } = useSelector(selector);

    const [page, setPage] = useState(1);
    const dispatch = useDispatch();

    const changePage = (nextPage) => {
        if(isLoading){
            return;
        }

        const currentData = data[nextPage - 1];
        if(currentData){
            dispatch(setCurrentProperties({ currentData, currentPage: nextPage }));
            return setPage(nextPage);
        }

        const pagesCount = nextPage - pages; 
        dispatch(getProperties({ ...query, pages: pagesCount }))
        setPage(nextPage)
    }

    useEffect(() => {
        setPage(1);
    }, [query.name])

    return(
        <> {maxPages > 0 &&
            <div>
                {page > 1 &&
                    <button data-testid='back' onClick={() => changePage(page - 1)}>prev</button>
                }
                <ul>
                    {
                        Array.from({length: page / pagesPerSlide < 1 ? pagesPerSlide : pagesPerSlide + 1 }).map((el, i) => {
                            const slide = Math.floor(page / pagesPerSlide);
                            let pageIndex = slide * pagesPerSlide + i;
                            pageIndex += slide == 0 ? 1 : 0;

                            if(pageIndex <= maxPages){
                                return <li $isSelected={pageIndex == page} key={pageIndex} data-testid={`${pageIndex}`} onClick={() => changePage(pageIndex)}>{pageIndex}</li>}
                            }
                        )
                    }
                </ul>
                {page < maxPages &&
                    <button data-testid='next' onClick={() => changePage(page + 1)}>next</button>
                }
            </div>
        } </>
    )
}
export default PropertiesPagination