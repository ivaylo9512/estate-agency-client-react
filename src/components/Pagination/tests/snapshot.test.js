import React from 'react';
import { shallow } from 'enzyme';
import Pagination from 'components/Pagination/PropertiesPagination';
import * as redux from 'react-redux';

describe('Pagination snpashot tests', () => {
    let selectorMock;

    beforeEach(() => {
        selectorMock = jest.spyOn(redux, 'useSelector');
        jest.spyOn(redux, 'useDispatch').mockReturnValue(jest.fn());
    })

    const createWrapper = (state) => {
        selectorMock.mockReturnValue(state);

        return shallow(<Pagination getProperties={jest.fn()} pagesPerSlide={5}/>)
    }

    it('renders correctly with pages', () => {
        const wrapper = createWrapper({ dataInfo: { data: [], maxPages: 5, isLoading: false}, query: { name: '', take: '10'}});

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly without pages', () => {
        const wrapper = createWrapper({ dataInfo: { data: [], maxPages: 0, isLoading: false}, query: { name: '', take: '10'}});

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly with pages at page 2 and back button', () => {
        const wrapper = createWrapper({ dataInfo: { data: [], maxPages: 5, isLoading: false}, query: { name: '', take: '10'}});
        wrapper.findByTestid(2).simulate('click');

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly with page that is last page of slide should render next slide', () => {
        const wrapper = createWrapper({ dataInfo: { data: [], maxPages: 10, isLoading: false}, query: { name: '', take: '10'}});
        wrapper.findByTestid(5).simulate('click');

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly with pages at last page', () => {
        const wrapper = createWrapper({ dataInfo: { data: [], maxPages: 4, isLoading: false}, query: { name: '', take: '10'}});
        wrapper.findByTestid(4).simulate('click');

        expect(wrapper).toMatchSnapshot();
    })
})