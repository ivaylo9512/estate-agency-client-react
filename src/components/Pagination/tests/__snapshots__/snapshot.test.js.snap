// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination snpashot tests renders correctly with page that is last page of slide should render next slide 1`] = `
<Fragment>
   
  <div>
    <button
      data-testid="back"
      onClick={[Function]}
    >
      prev
    </button>
    <ul>
      <li
        $isSelected={true}
        data-testid="5"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
      <li
        $isSelected={false}
        data-testid="6"
        key="6"
        onClick={[Function]}
      >
        6
      </li>
      <li
        $isSelected={false}
        data-testid="7"
        key="7"
        onClick={[Function]}
      >
        7
      </li>
      <li
        $isSelected={false}
        data-testid="8"
        key="8"
        onClick={[Function]}
      >
        8
      </li>
      <li
        $isSelected={false}
        data-testid="9"
        key="9"
        onClick={[Function]}
      >
        9
      </li>
      <li
        $isSelected={false}
        data-testid="10"
        key="10"
        onClick={[Function]}
      >
        10
      </li>
    </ul>
    <button
      data-testid="next"
      onClick={[Function]}
    >
      next
    </button>
  </div>
   
</Fragment>
`;

exports[`Pagination snpashot tests renders correctly with pages 1`] = `
<Fragment>
   
  <div>
    <ul>
      <li
        $isSelected={true}
        data-testid="1"
        key="1"
        onClick={[Function]}
      >
        1
      </li>
      <li
        $isSelected={false}
        data-testid="2"
        key="2"
        onClick={[Function]}
      >
        2
      </li>
      <li
        $isSelected={false}
        data-testid="3"
        key="3"
        onClick={[Function]}
      >
        3
      </li>
      <li
        $isSelected={false}
        data-testid="4"
        key="4"
        onClick={[Function]}
      >
        4
      </li>
      <li
        $isSelected={false}
        data-testid="5"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
    </ul>
    <button
      data-testid="next"
      onClick={[Function]}
    >
      next
    </button>
  </div>
   
</Fragment>
`;

exports[`Pagination snpashot tests renders correctly with pages at last page 1`] = `
<Fragment>
   
  <div>
    <button
      data-testid="back"
      onClick={[Function]}
    >
      prev
    </button>
    <ul>
      <li
        $isSelected={false}
        data-testid="1"
        key="1"
        onClick={[Function]}
      >
        1
      </li>
      <li
        $isSelected={false}
        data-testid="2"
        key="2"
        onClick={[Function]}
      >
        2
      </li>
      <li
        $isSelected={false}
        data-testid="3"
        key="3"
        onClick={[Function]}
      >
        3
      </li>
      <li
        $isSelected={true}
        data-testid="4"
        key="4"
        onClick={[Function]}
      >
        4
      </li>
    </ul>
  </div>
   
</Fragment>
`;

exports[`Pagination snpashot tests renders correctly with pages at page 2 and back button 1`] = `
<Fragment>
   
  <div>
    <button
      data-testid="back"
      onClick={[Function]}
    >
      prev
    </button>
    <ul>
      <li
        $isSelected={false}
        data-testid="1"
        key="1"
        onClick={[Function]}
      >
        1
      </li>
      <li
        $isSelected={true}
        data-testid="2"
        key="2"
        onClick={[Function]}
      >
        2
      </li>
      <li
        $isSelected={false}
        data-testid="3"
        key="3"
        onClick={[Function]}
      >
        3
      </li>
      <li
        $isSelected={false}
        data-testid="4"
        key="4"
        onClick={[Function]}
      >
        4
      </li>
      <li
        $isSelected={false}
        data-testid="5"
        key="5"
        onClick={[Function]}
      >
        5
      </li>
    </ul>
    <button
      data-testid="next"
      onClick={[Function]}
    >
      next
    </button>
  </div>
   
</Fragment>
`;

exports[`Pagination snpashot tests renders correctly without pages 1`] = `
<Fragment>
   
   
</Fragment>
`;
